### PR TITLE
Refactor `cv.entry` to match new metadata scheme (see #93)

### PR DESCRIPTION
**Pull request by DunklesArchipel**
The code was adapted to the new metadata scheme suggested in https://github.com/echemdb/svgdigitizer/issues/106 and implemented in https://github.com/echemdb/svgdigitizer/issues/112.

The new metadata scheme has an entry `figure_description` with original figure properties and an entry `data_description` describing the data in the CSV file (usually in SI units).

This way `echemdb` can also be used for non-digitized raw data, for which the metadata does not contain a `figure_description` but only contains a `data_description`.

### CODE REVIEW

**Comment by DunklesArchipel** [comment]
@saraedum @linuxrider this PR is now ready for review and should be merged before #102. I will try to fix the latter in the meanwhile.

File: echemdb/data/cv/descriptor.py [hunk 1]
```diff
@@ -104,10 +104,12 @@ def __getattr__(self, name):
 
         """
         name = name.replace('_', ' ')
-        if name in self._descriptor:
-            return Descriptor(self._descriptor[name])
-
-        raise AttributeError(f"Descriptor has no entry {name}. Did you mean one of {[key.replace(' ', '_') for key in self._descriptor.keys()]}?")
+
```
Comment (saraedum) [inline echemdb/data/cv/descriptor.py:107]: I  think you want to revert this change. This should be fixed by now (and returning `None` here is not alowed.)

File: echemdb/data/cv/descriptor.py [hunk 2]
```diff
@@ -120,10 +122,11 @@ def __getitem__(self, name):
             0
 
         """
-        if name in self._descriptor:
-            return Descriptor(self._descriptor[name])
-
-        raise KeyError(f"Descriptor has no entry {name}. Did you mean one of {list(self._descriptor.keys())}?")
+        try:
```
Comment (saraedum) [inline echemdb/data/cv/descriptor.py:125]: Same here. This should be reverted.

File: echemdb/data/cv/entry.py [hunk 3]
```diff
@@ -175,29 +175,29 @@ def format_title(self, e, which_field, as_sentence=True):
 
     def x(self):
         r"""
-        Return the name of the variable on the x-axis, i.e., `"U"`.
-
-        TODO: Adapt along with https://github.com/echemdb/svgdigitizer/issues/106.
+        Return the name of the variable on the x-axis, i.e., `"E"`.
```
Comment (saraedum) [inline echemdb/data/cv/entry.py:178]:
```suggestion
        Return the name of the variable on the x-axis, i.e., ``"E"``.
```

File: echemdb/data/cv/entry.py [hunk 4]
```diff
@@ -175,29 +175,29 @@ def format_title(self, e, which_field, as_sentence=True):
...
-            'U'
+            'E'
 
         """
         from astropy import units as u
-        if u.Unit(self.figure_description.potential_scale.unit).is_equivalent('V'):
-            return 'U'
+
+        if not self.data_description.axes:
+            raise ValueError(f"No axes were specified for the dataset.")
```
Comment (saraedum) [inline echemdb/data/cv/entry.py:190]: I think this is not the appropriate exception here since there is no argument. From the Python docs:
```
Raised when an operation or function receives an argument that has the right type but an inappropriate value, and the situation is not described by a more precise exception such as IndexError.
```

File: echemdb/data/cv/entry.py [hunk 5]
```diff
@@ -175,29 +175,29 @@ def format_title(self, e, which_field, as_sentence=True):
...
+        if 'E' in self.data_description.axes._descriptor.keys():
+            assert u.Unit(self.data_description.axes.E.unit).is_equivalent('V'), f"The variable on the x-axis is not equivalent to 'V'."
```
Comment (saraedum) [inline echemdb/data/cv/entry.py:192]:
```suggestion
            assert u.Unit(self.data_description.axes.E.unit).is_equivalent('V'), f"The variable on the x-axis is not convertible to 'V'."
```

I don't think people understand what they mean with "equivalent" here.

File: echemdb/data/cv/entry.py [hunk 6]
```diff
@@ -175,29 +175,29 @@ def format_title(self, e, which_field, as_sentence=True):
...
             >>> entry.x()
-            'U'
+            'E'
 
         """
         from astropy import units as u
-        if u.Unit(self.figure_description.potential_scale.unit).is_equivalent('V'):
-            return 'U'
+
+        if not self.data_description.axes:
```
Comment (saraedum) [inline echemdb/data/cv/entry.py:189]: I think this implementation is more complex than it has to be. The more straightforward implementation would be:

```
if not axes:
    raise
if 'E' not in:
    raise
if not equivalent:
    raise # or why would you want to only assert this?

return 'E'
```

File: echemdb/data/cv/entry.py [hunk 7]
```diff
@@ -207,12 +207,14 @@ def y(self):
         """
         from astropy import units as u
 
-        if u.Unit(self.figure_description.current.unit).is_equivalent('A / m2'):
-            return 'j' 
-        if u.Unit(self.figure_description.current.unit).is_equivalent('A'):
+        if self.data_description.axes.I:
```
Comment (saraedum) [inline echemdb/data/cv/entry.py:210]: Same here, I think the implementation can be simplified.

File: echemdb/data/cv/entry.py [hunk 8]
```diff
@@ -348,14 +351,28 @@ def __repr__(self):
...
             >>> entry.plot()
             Figure(...)
 
+        The plot can also be returned with the axis units of the original figure::
+
+            >>> entry = Entry.create_examples()[0]
+            >>> entry.plot(xunit='original', yunit='original')
+            Figure(...)
+
+        The plot can also be returned with custum axis units, where
```
Comment (saraedum) [inline echemdb/data/cv/entry.py:368]:
```suggestion
        The plot can also be returned with custom axis units, where 
```

File: echemdb/data/cv/entry.py [hunk 9]
```diff
@@ -348,14 +351,28 @@ def __repr__(self):
...
             Figure(...)
 
+        The plot can also be returned with the axis units of the original figure::
+
+            >>> entry = Entry.create_examples()[0]
+            >>> entry.plot(xunit='original', yunit='original')
+            Figure(...)
+
+        The plot can also be returned with custum axis units, where 
+        `xunit` should be equivalents to `V` and
```
Comment (saraedum) [inline echemdb/data/cv/entry.py:369]:
```suggestion
        `xunit` should be equivalent to `V` and 
```

File: echemdb/data/cv/entry.py [hunk 10]
```diff
@@ -348,14 +351,28 @@ def __repr__(self):
...
 
+        The plot can also be returned with the axis units of the original figure::
+
+            >>> entry = Entry.create_examples()[0]
+            >>> entry.plot(xunit='original', yunit='original')
+            Figure(...)
+
+        The plot can also be returned with custum axis units, where 
+        `xunit` should be equivalents to `V` and 
+        yunit equivalnts to `A` or `A / m2`.::
```
Comment (saraedum) [inline echemdb/data/cv/entry.py:370]:
```suggestion
        yunit equivalent to `A` or `A / m2`.::
```
:rofl:

**Review by saraedum — APPROVED** [review]
Some minor things could be improved. Otherwise, looks good.

File: echemdb/data/cv/entry.py (hunk 5 above)
Comment (DunklesArchipel) [reply echemdb/data/cv/entry.py:192]: agreed

File: echemdb/data/cv/descriptor.py (hunk 1 above)
Comment (DunklesArchipel) [reply echemdb/data/cv/descriptor.py:107]: my mistake. The old PR was merged in this one too.

File: echemdb/data/cv/entry.py (hunk 6 above)
Comment (DunklesArchipel) [reply echemdb/data/cv/entry.py:189]: @saraedum I hope this is what you wanted, as well as for `y` down below.

Comment (DunklesArchipel) [reply echemdb/data/cv/entry.py:189]: I guess you can merge this PR if you agree.

File: echemdb/data/cv/entry.py [hunk 11]
```diff
@@ -207,12 +208,20 @@ def y(self):
...
         from astropy import units as u
 
-        if u.Unit(self.figure_description.current.unit).is_equivalent('A / m2'):
-            return 'j' 
-        if u.Unit(self.figure_description.current.unit).is_equivalent('A'):
-            return 'I'
-        else:
-            raise ValueError(f"The variable on the y-axis is not equivalent to 'A / m2' or 'A'.")
+        if not self.data_description.axes:
+            raise IndexError("No axes were specified for the dataset.")
```
Comment (saraedum) [inline echemdb/data/cv/entry.py:212]:
```suggestion
            raise Exception("No axes were specified for the dataset.")
```

An index error should be raised when something (e.g. a list) is accessed out-of-bounds. If in doubt, just use `Exception`.

File: echemdb/data/cv/entry.py [hunk 12]
```diff
@@ -207,12 +208,20 @@ def y(self):
...
+        if not 'I' or not 'j' in self.data_description.axes._descriptor.keys():
+            raise KeyError("None of the axes has a variable 'I' or 'j'.")
+        if {'j', 'I'} <= self.data_description.axes._descriptor.keys():
+            raise KeyError("The axes should only contain the variable 'I' or 'j', but not both.")
```
Comment (saraedum) [inline echemdb/data/cv/entry.py:216]: This won't work. The condition evaluates as `not 'I'` (which is False because `'I'` is True) `or` the other part.

```suggestion
        if len({'j', 'I'}.intersection(self.data_description.axes._descriptor.keys())) == 1:
            raise KeyError("Exactly one axis must be labeled 'I' or 'j'.")
```

File: echemdb/data/cv/entry.py [hunk 13]
```diff
@@ -207,12 +208,20 @@ def y(self):
         """
         from astropy import units as u
 
-        if u.Unit(self.figure_description.current.unit).is_equivalent('A / m2'):
-            return 'j' 
-        if u.Unit(self.figure_description.current.unit).is_equivalent('A'):
-            return 'I'
-        else:
-            raise ValueError(f"The variable on the y-axis is not equivalent to 'A / m2' or 'A'.")
+        if not self.data_description.axes:
```
Comment (saraedum) [inline echemdb/data/cv/entry.py:211]: I don't think this works. When no axes are specified at all, this will raise an exception.

File: echemdb/data/cv/entry.py [hunk 14]
```diff
@@ -175,28 +175,29 @@ def format_title(self, e, which_field, as_sentence=True):
...
+            raise IndexError("No axes were specified for the dataset.")
+        if 'E' not in self.data_description.axes._descriptor.keys():
+            raise KeyError("None of the axes has a variable 'E'.")
+        if not u.Unit(self.data_description.axes.E.unit).is_equivalent('V'):
+            raise ValueError("The variable on the x-axis denoted as'E', is not convertible to 'V'.")
+        
+        return 'E'
 
     def y(self):
```
Comment (saraedum) [inline echemdb/data/cv/entry.py:198]: I think technically, you are missing the case here that the axis is labeled `I` but in `A / m²`.

File: echemdb/data/cv/entry.py [hunk 15]
```diff
@@ -175,28 +175,27 @@ def format_title(self, e, which_field, as_sentence=True):
...
+            'E'
 
         """
         from astropy import units as u
-        if u.Unit(self.figure_description.potential_scale.unit).is_equivalent('V'):
-            return 'U'
-        else:
-            raise ValueError(f"The variable on the x-axis is not equivalent to 'V'.")
+
+        axis = getattr(self.data_description, 'axes', {})['E']
```
Comment (saraedum) [inline echemdb/data/cv/entry.py:189]: This now throws a KeyError when there are no axes and when there is no E axis. In both cases, the error says that there is no E which is what we want here I think.

File: echemdb/data/cv/entry.py [hunk 16]
```diff
@@ -207,12 +206,24 @@ def y(self):
...
+
+        variable = variables[0]
+
+        if variable == 'I':
+            if not u.Unit(axes[variable].unit).is_equivalent('A'):
+                raise Exception("Unit on I axis must be convertible to A.")
+
+        if variable == 'j':
+            if not u.Unit(axes[variable].unit).is_equivalent('A / m2'):
+                raise Exception("Unit on I axis must be convertible to A / m².")
```
Comment (saraedum) [inline echemdb/data/cv/entry.py:224]:
```suggestion
                raise Exception("Unit on j axis must be convertible to A / m².")
```

**Comment by saraedum** [comment]
@DunklesArchipel I fixed a minor bug and tried to simplify things a bit. Please have a look and merge if this makes sense to you.
